articles/actions.md

```diff
@@ -0,0 +1,338 @@
+
+State transitions triggered by the action client:
+
+- **send_goal** - A goal is sent to the action server.
+The state machine is only started if the action server *accepts* the goal.
+- **cancel_goal** - Request that the action server stop processing the goal.
+A transition only occurs if the action server *accepts* the request to cancel the goal.
+The goal may transition to CANCELING or CANCELED depending on the action server implementation.
+
+## API
+
+Proposed examples can be in the respository [examples (branch: actions_proposal)](https://github.com/ros2/examples/tree/actions_proposal).
+
+C++:
+
+- [examples/rclcpp/minimal_action_server](https://github.com/ros2/examples/tree/actions_proposal/rclcpp/minimal_action_server)
+- [examples/rclcpp/minimal_action_client](https://github.com/ros2/examples/tree/actions_proposal/rclcpp/minimal_action_client)
+
+Python:
+
+- [examples/rclpy/actions](https://github.com/ros2/examples/tree/actions_proposal/rclpy/actions)
+
+## Middleware implementation
+
+Under the hood, an action is made up of three services and two topics, each descibed in detail here.
+
+![Action Client/Server Interaction Overview](../img/actions/interaction_overview.png)
+
+### Goal Submission Service
+
+* **Direction**: Client calls Server
+* **Request**: Description of goal
+* **Response**: Whether goal was accepted or rejected, a unique identifier for the goal, and the time when the goal was accepted.
+
+The purpose of this service is to submit a goal to the action server.
+It is the first service called to begin an action, and is expected to return quickly.
+A user-define description of the goal is sent as the request.
+The response indicates whether or not the goal was accepted, and if so the identifier the server will use to describe the goal.
+
+The QoS settings of this service should be set the so the client is guaranteed to receive a response or an action could be executed without a client being aware of it.
+
+### Cancel Request Service
+
+* **Direction**: Client calls Server
+* **Request**: Goal identifier, time stamp
+* **Response**: Goals that will be attempted to be canceled
+
+The purpose of this service is to request to cancel one or more goals on the action server.
+A cancellation request may cancel multiple goals.
+The result indicates which goals will be attempted to be canceled.
+Whether or not a goal is actually canceled is indicated by the status topic and the result service.
+
+The cancel request policy is the same as in ROS 1.
+
+* If the goal ID is empty and time is zero, cancel all goals
+* If the goal ID is empty and time is not zero, cancel all goals accepted at or before the time stamp
+* If the goal ID is not empty and time is not zero, cancel the goal with the given id regardless of the time it was accepted
+* If the goal ID is not empty and time is zero, cancel the goal with the given id and all goals accepted at or before the time stamp
+
+### Get Result Service
```

Comment:
We could have the option to cache the result for a longer time in the action server (configurable timeout?) so multiple clients could access it, like command line tools. But perhaps there is a case to be made for a topic instead of a service.
In any case, I think we should have a tool like `ros2 action result` for introspection. I will add it to the introspection tools section.

Reply:
I believe feature for caching results and retrieving with a CLI tool is covered in this doc.